internal/environments/environments.go

```diff
@@ -0,0 +1,110 @@
+package environments
+
+import (
+	"github.com/trento-project/trento/internal/consul"
+	"github.com/trento-project/trento/internal/hosts"
+)
+
+type Environment struct {
+	Name       string                `mapstructure:"name,omitempty"`
+	Landscapes map[string]*Landscape `mapstructure:"landscapes,omitempty"`
+}
+
+type Landscape struct {
+	Name       string                `mapstructure:"name,omitempty"`
+	SAPSystems map[string]*SAPSystem `mapstructure:"sapsystems,omitempty"`
+}
+
+type SAPSystem struct {
+	Name  string         `mapstructure:"name,omitempty"`
+	Type  string         `mapstructure:"type,omitempty"`
+	Hosts hosts.HostList `mapstructure:"hosts,omitempty"`
+}
+
+func NewEnvironment(env, land, sys string) Environment {
+	newSys := &SAPSystem{Name: sys}
+	systemsMap := make(map[string]*SAPSystem)
+	systemsMap[sys] = newSys
+
+	newLand := &Landscape{Name: land, SAPSystems: systemsMap}
+	landsMap := make(map[string]*Landscape)
+	landsMap[land] = newLand
+
+	newEnv := Environment{Name: env, Landscapes: landsMap}
+	return newEnv
+}
+
+func ungrouped(name string) bool {
+	return name == consul.KvUngrouped
+}
+
+func (e *Environment) Ungrouped() bool {
+	return ungrouped(e.Name)
+}
+
+func (l *Landscape) Ungrouped() bool {
+	return ungrouped(l.Name)
+}
+
+func (s *SAPSystem) Ungrouped() bool {
+	return ungrouped(s.Name)
+}
+
+type EnvironmentHealth struct {
+	Health    string
+	HealthMap map[string]string
+}
+
+func (e *EnvironmentHealth) updateHealth(n string, h string) EnvironmentHealth {
+	e.HealthMap[n] = h
+
+	if h == "critical" {
+		e.Health = h
+	} else if h == "warning" && e.Health != "critical" {
+		e.Health = h
+	}
+
+	return *e
+}
+
+func (e *Environment) Health() EnvironmentHealth {
+	var health = EnvironmentHealth{
+		Health:    "passing",
+		HealthMap: make(map[string]string),
+	}
+
+	for _, land := range e.Landscapes {
+		h := land.Health().Health
+		health = health.updateHealth(land.Name, h)
+	}
+
+	return health
+}
+
+func (l *Landscape) Health() EnvironmentHealth {
+	var health = EnvironmentHealth{
+		Health:    "passing",
+		HealthMap: make(map[string]string),
+	}
+
+	for _, system := range l.SAPSystems {
+		h := system.Health().Health
+		health = health.updateHealth(system.Name, h)
+	}
+
+	return health
+}
+
+func (s *SAPSystem) Health() EnvironmentHealth {
+	var health = EnvironmentHealth{
+		Health:    "passing",
+		HealthMap: make(map[string]string),
+	}
+
+	for _, host := range s.Hosts {
+		h := host.Health()
+		health = health.updateHealth(host.Name(), h)
+	}
+
+	return health
+}
```

Comment:
There is an extreme degree of repetition here. Have you thought about generalizing getting the name of an entity and their health status into some interfaces?

Also, what happens if there is a clash between host names, system names, and landscape names? We are using the same map for all of these three, and I think this might potentially cause problems.

Reply:
Actually, I just copy/pasted already existing code, so I didn't even have a look. Anyway, I did myself this code in the past. 
I will try to make it more reusable

Reply:
About the last comment about a potential clash, you are right. I didn't consider this scenario. I will need to think about it, maybe having some kind of nested map or similar. Let's see what I can think of

Reply:
Hey @stefanotorresi ,
I have done some changes in https://github.com/trento-project/trento/pull/58/commits/8932b53821e42ee50311ac2007c5663691622565